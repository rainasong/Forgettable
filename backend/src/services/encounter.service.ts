/* eslint-disable no-underscore-dangle */
import mongoose from 'mongoose';
import Encounter, { EncounterModel } from '../models/encounter.model';
import logger from '../utils/logger';

const algoliaSearch = require('algoliasearch');

const client = algoliaSearch(
  process.env.ALGOLIA_APP_ID,
  process.env.ALGOLIA_SECRET_KEY,
);
const index = client.initIndex('encounters');

const createEncounter = async (encounterDetails: EncounterModel) => {
  const encounter = new Encounter(encounterDetails);

  await encounter.save();

  if (encounter.persons.length === 0) {
    await Encounter.deleteOne({ _id: encounter._id }).exec();
    const e = new Error('Persons can\'t be empty');
    e.name = 'ValidationError';
    throw e;
  }

  return encounter;
};

const getEncounter = async (encounterId) => Encounter.findOne({_id: encounterId});

const getAllEncounters = async (queryParams: any, userEncounters: mongoose.Types.ObjectId[]) => {
  
  // Get all encounters from the db which belongs to the user
  let foundUserEncounters = await Encounter.find({ _id: { $in: userEncounters } });

  // Filter the found encounters by query params
  if (queryParams.term) {
    logger.info('Query params received:');
    logger.info(queryParams);
    const termValue = queryParams.term.toLowerCase();

    const algoliaSearchResults = await index.search(termValue);
    const algoliaSearchResultsIds = algoliaSearchResults.hits.map((hit) => hit.objectID.toString())

    const userEncounterResults = foundUserEncounters.filter(
      (encounter) => algoliaSearchResultsIds.includes(encounter._id.toString()),
    );

    foundUserEncounters = userEncounterResults;
  }

  return foundUserEncounters;
};

const updateEncounter = async (objectID: string, encounterDetails: EncounterModel) => {
  console.log(objectID);
  const updatedEncounter = await Encounter
    .findByIdAndUpdate(objectID, encounterDetails, { new: true });

  const updatedEncounterAlgolia : any = await Encounter.findById(objectID);
  updatedEncounterAlgolia.objectID = objectID;
  await index.partialUpdateObject(updatedEncounterAlgolia);

  return updatedEncounter;
};

// Service is used for deletePersons endpoint
const deleteEncounterPerson = async (personID: string) => {
  // Remove the person with the supplied ID from all existing encounters
  const result = await Encounter.updateMany({ }, { $pullAll: {persons: [{ _id: personID}]} });
  // Store references to the encounters that no longer contain persons so they can be removed from Users later
  const empty_encounters = await Encounter.find({persons : {$exists:true, $size:0}});
  // Delete all the empty encounters
  const deleted_encounters = await Encounter.deleteMany({persons: {$exists: true, $size: 0}});

  // Check that all deleted encounters are returned and Encounters have been updated 
  if (empty_encounters.length == deleted_encounters.deletedCount) {
    return {"array": empty_encounters, "bool": true};
  } else {
    return {"array": empty_encounters, "bool": false};
  }
}

const deleteEncounter = async (encounterID: String) => {
  const result = await Encounter.deleteOne({ _id: encounterID }).exec();
  
  // Check that Encounter has been deleted
  if (result.deletedCount == 1) {
    await index.deleteObject(encounterID);
    return true;
  } else {
    return false;
  }

};

const encounterService = {
  createEncounter,
  updateEncounter,
  getEncounter,
  getAllEncounters,
  deleteEncounter,
  deleteEncounterPerson,
};

export default encounterService;
