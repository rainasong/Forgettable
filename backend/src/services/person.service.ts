/**
 * Service contains database operations
 */
import mongoose from 'mongoose';
import Person, { PersonModel } from '../models/person.model';
import logger from '../utils/logger';

const algoliaSearch = require('algoliasearch');

const client = algoliaSearch(
  process.env.ALGOLIA_APP_ID,
  process.env.ALGOLIA_SECRET_KEY,
);
const index = client.initIndex('persons');

const createPerson = async (personDetails: PersonModel) => {
  const person = new Person(personDetails);
  await person.save();
  return person;
};

const updatePersonWithId = async (reqPersonId: string, personNewDetails: PersonModel) => {
  const query = { _id: reqPersonId };
  
  const updatedPerson = await Person.findOneAndUpdate(query, personNewDetails, { upsert: true });

  const updatedPersonAlgolia : any = await Person.findById(reqPersonId);
  updatedPersonAlgolia.objectID = reqPersonId;
  await index.partialUpdateObject(updatedPersonAlgolia);

  return updatedPerson;
};

const getPersonWithId = async (reqPersonId: string) => {
  const query = { _id: reqPersonId };
  return Person.findOne(query);
};

const getPeople = async (queryParams: any, userPersons: mongoose.Types.ObjectId[]) => {
  // Get all persons from the db that belong to the user
  let foundUserPersons = await Person.find({ _id: { $in: userPersons } });

  // Filter the found persons by query params
  if (queryParams.term) {
    logger.info('Query params received:');
    logger.info(queryParams);
    const termValue = queryParams.term.toLowerCase();

    const algoliaSearchResults = await index.search(termValue);
    const algoliaSearchResultsIds = algoliaSearchResults.hits.map((hit) => hit.objectID.toString())
  
    const userPersonResults = foundUserPersons.filter(
      (person) => algoliaSearchResultsIds.includes(person._id.toString()),
    );

    foundUserPersons = userPersonResults;
  }

  return foundUserPersons;
};

const deletePersonEncounters = async (encounterID: string) => {
  const result = await Person.updateMany({ }, { $pullAll: {encounters: [{ _id: encounterID}]} });

  // Check that Persons with the respective encounters has been updated
  if (result.modifiedCount > 0) {
    return true;
  } else {
    return false;
  }
}

const deletePersons = async (personID: string) => {
  const result = await Person.deleteOne({_id: personID});

  if (result.deletedCount == 1) {
    await index.deleteObject(personID);
    return true;
  } else {
    return false;
  }
}

const addEncounterToPersons = async (personIds, encounterId) => {
  for (let i = 0; i < personIds.length; i++) {
    const result = await Person
      .updateOne({ _id: personIds[i] }, { $push: { encounters: encounterId } });

    // Revert all updates if any update fails
    if (result.modifiedCount !== 1) {
      for (let j = i - 1; j >= 0; j--) {
        await Person.updateOne({ _id: personIds[i] }, { $pop: { encounters: 1 } });
      }
      return false;
    }
  }

  return true;
};

const personService = {
  createPerson,
  getPersonWithId,
  getPeople,
  deletePersonEncounters,
  deletePersons,
  addEncounterToPersons,
  updatePersonWithId,
};

export default personService;
