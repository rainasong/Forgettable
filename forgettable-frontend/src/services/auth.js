import firebase from 'firebase/compat/app';
import 'firebase/compat/auth';
import firebaseConfig from '../firebase-config';
import 'regenerator-runtime/runtime';
import {getAuth, signInWithPopup} from 'firebase/auth';

/**
 * Contains logic for handling everything auth related. This includes
 * logging in, logging out, and persisting login status.
 * Stores the user's name, email and other meta info in localstorage.
 * JWT is NOT stored.
 *
 * Author: Mercury Lin (lin8231)
 */
const app = firebase.initializeApp(firebaseConfig);
export const authentication = getAuth(app);

const googleProvider = new firebase.auth.GoogleAuthProvider();
const signIn = (callback) => {
  signInWithPopup(authentication, googleProvider)
      .then((result) => {
        const user = result.user;

        if (user) {
          callback(true, {
            userName: user.displayName,
            userId: user.uid,
          });
        }
      })
      .catch((error) => {
        console.log('Uh oh, error loggin in :( ', error);
        callback(false);
      });
};

const signOut = () => {
  firebase.auth().signOut();
};


const persistLoginStatus = (user) => {
  // save user details to localstorage
  localStorage.setItem('user', JSON.stringify(user));
};

const loadLoginStatus = () => {
  // read user details from localstorage
  const user = JSON.parse(localStorage.getItem('user'));
  if (user) {
    return {
      isLoggedIn: true,
      userInfo: user,
    };
  } else {
    return {
      isLoggedIn: false,
    };
  }
};

const clearPersistedLoginStatus = () => {
  // clear user details from localstorage
  localStorage.removeItem('user');
};

export default {
  signIn: signIn,
  signOut: signOut,
  persistLoginStatus: persistLoginStatus,
  loadLoginStatus: loadLoginStatus,
  clearPersistedLoginStatus: clearPersistedLoginStatus,
};
