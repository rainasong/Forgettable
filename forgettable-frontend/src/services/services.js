import React from 'react';
import axios from 'axios';

export const getPerson = (personId) => {

};

export const getAllPersons = () => {

};

export const createPerson = (person) => {
  // {
  //     full_name: string,
  //     birthday?: Date,
  //     gender?: string,
  //     location?: string,
  //     first_met?: Date,
  //     how_we_met?: string,
  //     interests?: string[],
  //     organisation?: string,
  //     social_media?: Map<string>,
  //     image?: base64-image,
  //     encounters?: ObjectId[],
  //     time_updated: Date
  //   }

};

export const updatePerson = (person) => {

};

export const deletePerson = (personId) => {

};

export const getEncounter = (encounterId) => {

};

export const getAllEncounters = () => {

};

export const createEncounter = (encounter) => {
  // {
  //     date:? Date,
  //     location?: string,
  //     title: string,
  //     description?: string,
  //     persons?: [ObjectId]
  //   }
};

// export const updateEncounter = () => {
// do we want this?
// }

export const deleteEncounter = (encounterId) => {

};
