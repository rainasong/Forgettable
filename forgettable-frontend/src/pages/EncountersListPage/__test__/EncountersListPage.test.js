import React from 'react';
import {screen, render, fireEvent, waitFor} from '@testing-library/react';
import {act} from 'react-dom/test-utils';
import * as api from '../../../services';
import ShallowRenderer from 'react-test-renderer/shallow';
import EncountersListPage from '../EncountersListPage';

jest.mock('../../../services');

it('renders EncountersList UI with correct hierarchy', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<EncountersListPage/>);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
