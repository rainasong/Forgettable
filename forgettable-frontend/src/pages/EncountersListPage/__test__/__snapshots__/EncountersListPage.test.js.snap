// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EncountersList UI with correct hierarchy 1`] = `
<div>
  <CustomModal
    hasCancel={true}
    hasConfirm={true}
    onClose={[Function]}
    onConfirm={[Function]}
    open={false}
  >
    <div
      className="DeleteModal"
    >
      <h1>
        Warning
      </h1>
      <p>
        Are you sure you want to delete this encounter? You cannot undo this action.
      </p>
    </div>
  </CustomModal>
  <div
    className="Container"
  >
    <div
      className="Header"
    >
      Encounters
    </div>
    <div
      className="Utilities"
    >
      <SearchBar
        exportSearchString={[Function]}
        hasAutocomplete={false}
        placeholder="Search"
      />
      <div
        className="Button"
      >
        <Link
          style={
            Object {
              "textDecoration": "none",
            }
          }
          to={
            Object {
              "pathname": "/encounters/create",
            }
          }
        >
          <IconButton
            btnText="New Encounter"
            includeIcon={true}
            onClick={[Function]}
          />
        </Link>
      </div>
    </div>
    <h4>
      Loading...
    </h4>
  </div>
  <ToastContainer
    autoClose={5000}
    closeButton={[Function]}
    closeOnClick={true}
    draggable={true}
    draggableDirection="x"
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="bottom-center"
    role="alert"
    rtl={false}
    theme="light"
    transition={[Function]}
  />
</div>
`;
